Automattic/woocommerce-payments
# Make Tracks identity cookie cache compatible

Fixes #8162 

#### Changes proposed in this Pull Request

This PR addresses an issue with the `tk_ai` cookie and its impact on page caching. Previously, the `tk_ai` cookie was set on page load, which made pages uncacheable. This change defers the setting of this cookie until it's actually needed, thus improving cacheability. The solution involves fetching the Tracks identity by checking for the presence of the tk_ai cookie and setting it via an Ajax request if it's not present.

See p6q8Tx-4j4-p2#comment-8955 for a detailed explanation of the issue. 

**Key changes:**

- Removal of the immediate setting of the `tk_ai` cookie on page load.
- Introduction of `getTracksIdentity` function in `client/tracks/index.ts` to asynchronously fetch the Tracks identity.
- Modification of iframe sources in `woopay-connect-iframe.js` and `woopay-express-checkout-button.js` to include the `tracksUserIdentity` parameter only after the Tracks identity is obtained.
- Removal of duplicated attempts to set `tk_ai` cookie in `class-woopay-tracker.php`









#### Testing instructions




**Setup**
- Enable WooPay.
- Make sure store country is set to US in WooCommerce -> Settings -> General.

**Reproduce the issue in `develop`**
1. As a shopper, visit the site in an incognito winow.
2. Open dev tools and make sure the `tk_ai` cookie is not set.
3. Add a product to the cart and visit the cart page.
4. Notice that the `Set-Cookie` header in response headers of the initial page request.

<img width="1509" alt="SCR-20240214-kujk" src="https://github.com/Automattic/woocommerce-payments/assets/6216000/fb272205-dab0-42fc-bd45-0e703b900e87">

**Test the fix in this branch**
1. Redo the above steps and make sure the Set-Cookie header is absent in the initial page load request. `tk_ai` cookie will still be set via an Ajax request.
2. Remove the `tk_ai` cookie.
3. Visit a product page and make sure `Set-Cookie` header is absent on the response headers.
4. Remove the cookie once more.
5. Visit the checkout page and make sure `Set-Cookie` header is absent on the response headers.

**Regression tests**
* Test whether the Tracks User ID is transferred to WooPay in the first-party auth flow.
1. Add a breakpoint [here](https://github.com/Automattic/woopay/blob/be1a12b871b431d11db0d88d731d64a68abf50d2/src/Connect/ConnectPage.php#L85).
2. As a shopper, add a product to the cart and visit the cart page.
3. Above breakpoint should be reached. 
4. Make sure the `tracks_user_identity` field inside the `env_data` variable contains `_ui` and `_ut` entries.

* Test whether Tracks User ID is transferred to WooPay in the iframe flow
1. As a shopper, add a product to the cart and go to the checkout page.
2. Enter a WooPay registered email address.
3. Wait for the OTP iframe to show up.
4. Inspect the iframe and make sure that the `tracksUserIdentity` is passed on to the iframe as a URL parameter.

<img width="1505" alt="SCR-20240214-ljwo" src="https://github.com/Automattic/woocommerce-payments/assets/6216000/69e125a9-01ac-4eb9-a747-b37940cb7d9e">

* Test the Direct Checkout Flow
1. Make sure Direct Checkout feature is enabled for your store.
2. As a shopper add a product to the cart and visit the cart page.
3. Open dev tools and ensure the Direct Checkout iframe contains the `tracksUserIdentity`

<img width="1072" alt="SCR-20240214-llib" src="https://github.com/Automattic/woocommerce-payments/assets/6216000/311630ef-5b12-4d6e-a6d1-f3112f47e671">


-------------------


**Post merge**
